### ISSUE TITLE
custom_resources: AwsCustomResource cannot handle updates requiring replacement

### ISSUE THREAD

**Issue by rittneje** [aws/aws-cdk#17443]
### What is the problem?

We can use AwsCustomResource to map custom resource create/update/delete to an SDK call. However, there are updates that actually require a new resource to be created and the original one deleted.

For example, suppose we are using such a custom resource to manage IoT Core role aliases. The create action maps to createRoleAlias, the update action maps to updateRoleAlias, and the delete action maps to deleteRoleAlias. However, if we are specifically trying to update the role alias name itself, then we cannot just call updateRoleAlias, because under normal circumstances the new role alias does exist yet and must be created not updated.

Even if some mechanism were added to AwsCustomResource or AwsSdkCall to detect updates requiring replacement, there is still an issue. If the deployment were to fail, CloudFormation will attempt to rollback by issuing another update back to the original parameters. But this would also be detected as a replacement, which would issue another createRoleAlias, which would fail because the original role alias still exists. So now you'd need something akin to `ignore_error_codes_matching`, but only in the specific case of a rollback after an update requiring replacement.

### Reproduction Steps

Use AwsCustomResource to create an IoT Core role alias. Then try to update the role alias name.

### What did you expect to happen?

It should be able to handle updates requiring replacement, or the construct should be deprecated with a warning that it cannot be used in most cases.

### What actually happened?

It tried to update a non-existent role alias and failed.

### CDK CLI Version

1.130.0

### Node.js Version

16.13.0

### OS

Alpine 3.14

### Language

Python

### Language Version

3.9

**Comment by rix0rrr** [comment]
The correct solution would be to create a new role and return a different `PhysicalResourceId`. CloudFormation will treat this is a replacement and issue a delete for the original resource in the cleanup phase.

This is probably worth adding to the documentation of the module.

**Comment by rittneje** [comment]
@rix0rrr How exactly would that work though? As far as I can tell I can only configure it to make a particular SDK call, so I don't see a way to cause the update logic to sometimes call updateRoleAlias and sometimes call createRoleAlias.

**Comment by rittneje** [comment]
@rix0rrr Please re-add the "bug" label to this issue unless you have a more concrete workaround.

**Comment by gshpychka** [comment]
@rix0rrr this will not work, since `onUpdate` calls the `updateRoleAlias`, which cannot create a new resource. So `onUpdate` will fail.

**Comment by rittneje** [comment]
@rix0rrr Please re-add the bug label.

**Comment by gshpychka** [comment]
> @rix0rrr Please re-add the bug label.

This isn't a bug, it's a feature request.

**Comment by rittneje** [comment]
@gshpychka No, it's a bug. As soon as you use AwsCustomResource, you get stuck in a bad state where you literally cannot perform certain updates. Either it needs to be fixed to function properly, or a warning needs to be added to the documentation not to use it.

**Comment by gshpychka** [comment]
> @gshpychka No, it's a bug. As soon as you use AwsCustomResource, you get stuck in a bad state where you literally cannot perform certain updates. Either it needs to be fixed to function properly, or a warning needs to be added to the documentation not to use it.

No, it's not as soon as you use it, it's only when you use it with specific API actions. It was meant for the simplest cases, you can use a lambda backed custom resource for something more complex.

**Comment by rittneje** [comment]
@gshpychka It is broken for any resource in which the "create" and "update" SDK calls are different, which is many of them. Consequently, it is a serious design flaw that AwsCustomResource has `on_create` and `on_update` as separate parameters. Furthermore, it does not restrict the API calls you can make in any way, so it does nothing to prevent such "misuse". And in order to fix this situation, we would have to change the `ServiceToken` to refer to a custom lambda function, but CloudFormation does not allow you to update that. https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cfn-customresource.html#cfn-customresource-servicetoken

In short, developers who don't realize the inherent issues with AwsCustomResource are mislead into using it, and then there is no way to fix it. I consider that to be a very serious bug.

**Comment by rittneje** [comment]
Don’t close.

**Comment by rittneje** [comment]
ping @rix0rrr @peterwoodworth
